File: trunk/acce2/core/fileServer.py
from glob import glob
import os.path

class FileServer:
    def __init__(self):
        self.files = [filename.replace('\\','/') for filename in self.enumerateFiles('.')]
        
        self.filesByType = {}
        self.filesByName = {}
        
        for filename in self.files:
            self.filesByType.setdefault(self.getType(filename),[]).append(filename)
            self.filesByName[self.getName(filename)] = filename
        
    def enumerateFiles(self, path, pattern = None):
        files = filter(os.path.isfile,glob("%s/%s" % (path,pattern or "*")))
        for subdir in filter(os.path.isdir,glob("%s/*" % path)):
            files.extend(self.enumerateFiles(subdir))
        return files

    def getType(self, filename):
        if '.' in filename:
            return filename[filename.rfind('.')+1:]
        else:
            return ''
            
    def getName(self, filename):
        return filename[filename.rfind('/')+1:]
        
    def getFile(self, filename):
        if os.path.isfile(filename):
            return filename
        else:
            assert filename in self.filesByName, "unknown file %s" % filename
            return self.filesByName[filename]
        
    def isFile(self, filename):
        return os.path.isfile(filename) or filename in self.filesByName

fileServer = FileServer()

if __name__ == "__main__":
    assert fileServer.getFile("fileServer.py") == "./core/fileServer.py"

File: trunk/sdse/core/FileServer.cpp
#include "FileServer.h"

using namespace sdse;

bool FileServer::filesEnumerated = false;
StringMap FileServer::filesByName;
StringListMap FileServer::filesByType;
StringList FileServer::filePaths;

void FileServer::checkFilesEnumerated() {
    if (!filesEnumerated) {
        filesEnumerated = true;
        clear();
        enumerateFiles(boost::filesystem::path("."));
    }
}

void FileServer::clear() {
    filesByName.clear();
    filesByType.clear();
    filePaths.clear();
}

void FileServer::enumerateFiles(boost::filesystem::path path) {
    boost::filesystem::directory_iterator end;
    filePaths.push_back(path.string());
    for (boost::filesystem::directory_iterator i(path); i != end; ++i) {
        if (boost::filesystem::is_directory(i->status())) {
            enumerateFiles(i->path());
        } else {
            std::string filename = i->string(),
                ext = getExtension(filename),
                name = getName(filename);
                
            filesByName[name] = filename;
            
            if (filesByType.find(ext) != filesByType.end()) filesByType[ext] = StringList();
            filesByType[ext].push_back(filename);
        }
    }
}

std::string FileServer::getFile(std::string filename) {
    if (boost::filesystem::exists(boost::filesystem::path(filename)))
        return filename;
    checkFilesEnumerated();
    StringMap::iterator i = filesByName.find(filename);
    if (i != filesByName.end())
        return i->second;
    else
        return "";
}

StringList FileServer::getFilesOfType(std::string ext) {
    checkFilesEnumerated();
    StringListMap::iterator i = filesByType.find(ext);
    if (i != filesByType.end())
        return filesByType[ext];
    else
        return StringList();
}

StringList FileServer::getFilePaths() {
    return filePaths;
}

std::string FileServer::getExtension(std::string filename) {
    int pos = filename.rfind(".");
    if (pos != std::string::npos)
        return filename.substr(pos+1);
    else
        return "";
}

std::string FileServer::getName(std::string filename) {
    return filename.substr(filename.rfind("/")+1);
}
